src/simulations/set_generator.py

```diff
@@ -486,3 +506,87 @@ def __iter__(self):
+  I.e., freq ~ Poission(freq_rate) + 1.
+  """
+
+  @classmethod
+  def get_generator_factory_with_num_and_size(cls, universe_size, num_sets,
+                                              set_size, freq_rate_list,
+                                              freq_cap):
+
+    def f(random_state):
+      return cls(universe_size, list(_SetSizeGenerator(num_sets, set_size)),
+                 freq_rate_list, random_state, freq_cap)
+
+    return f
+
+  @classmethod
+  def get_generator_factory_with_set_size_list(cls, universe_size,
+                                               set_size_list, freq_rate_list,
+                                               freq_cap):
+
+    def f(random_state):
+      return cls(universe_size, set_size_list, freq_rate_list, random_state,
+                 freq_cap)
+
+    return f
+
+  def __init__(self, universe_size, set_size_list, freq_rate_list, random_state,
+               freq_cap=None):
+    """Create a homogeneous multiset generator.
+
+    Args:
+      universe_size: An integer value that specifies the size of the whole
+        universe from which the ids will be sampled.
+      set_size_list: A list containing the size of each set, aka, the number of
+        reached IDs.
+      freq_rate_list: A list of the same size as set_size_list, specifying the
+        non-negative freq_rate parameter of the shifted-Possion distribution.
+      random_state: a numpy random state.
+      freq_cap: An optional positive integer which represents the maximum number
+        of times an ID can be seen in the returned set. If not set, will not
+        apply this capping.
+
+    Raises: AssertionError when (1) set_size_list and freq_rate_list do not
+      have equal length, (2) elements of freq_rate_list are not all
+      non-negative, or (3) freq_cap is not None or positive.
+    """
+    assert len(set_size_list) == len(freq_rate_list), (
+        'set_size_list and freq_rate_list do not have equal length.')
+    assert all([freq_rate >= 0 for freq_rate in freq_rate_list]), (
+        'Elements of freq_rate_list should be non-negative.')
+    assert freq_cap is None or freq_cap > 0, (
+        'freq_cap should be None or positive.')
+    self.universe_size = universe_size
+    self.set_size_list = set_size_list
+    self.freq_rate_list = freq_rate_list
+    self.freq_cap = freq_cap
+    self.random_state = random_state
+    self.choice = choice_fast  # For simple mock in unit testing.
+
+  def __iter__(self):
+    for set_size, freq_rate in zip(self.set_size_list, self.freq_rate_list):
+      set_ids = self.choice(
+          self.universe_size, set_size, self.random_state)
+      freq_per_id = self.random_state.poisson(
+          lam=freq_rate, size=len(set_ids)) + 1
+      if self.freq_cap is not None:
+        freq_per_id = np.minimum(freq_per_id, self.freq_cap)
+      multiset_ids = []
+      for i, freq in zip(set_ids, freq_per_id):
+        multiset_ids += [i] * freq
+      yield multiset_ids
```

Comment:
Should we shuffle the multiset before returning? I think this will have an effect on how same-key-aggregator behaves. This could be worth discussing in more detail offline.

Reply:
Done.